### PR TITLE
Data streams service to handle modification requests

### PR DESCRIPTION
**Pull request by danhermann**
Relates to #75852 and #77778

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDataStreamsService.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    private static void removeBackingIndex(Metadata metadata, Metadata.Builder builder, String dataStreamName, String indexName) {
+        var dataStream = validateDataStream(metadata, dataStreamName);
+        var index = validateIndex(metadata, indexName);
+
+        builder.put(dataStream.getDataStream().removeBackingIndex(index.getWriteIndex().getIndex()));
+
+        // remove or disable DataStreamTimestampFieldMapper
+        // TODO: is this necessary or desirable?
```
Comment (danhermann) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDataStreamsService.java:133]: @martijnvg, do you think we should either remove or disable the `DataStreamTimestampFieldMapper` on indices that are removed from data streams?

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDataStreamsService.java:133]: I think it is ok to keep this meta field as is?  If an index is removed from a data stream then this means that they always need to be providing a valid `@timestamp` field and value for new documents. Which kind of makes sense given the type of data that data streams hold.

File: server/src/test/java/org/elasticsearch/cluster/metadata/MetadataDataStreamsServiceTests.java [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+        assertThat(e.getMessage(), equalTo("index [" + missingIndex + "] not found"));
+    }
+
+    private MapperService getMapperService(IndexMetadata im) {
+        try {
+            return createMapperService("{\"_doc\": " + im.mapping().source().toString() + "}");
+        } catch (IOException e) {
+            throw new IllegalStateException(e);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (martijnvg) [reply server/src/test/java/org/elasticsearch/cluster/metadata/MetadataDataStreamsServiceTests.java:354]: Do you know what exception is thrown when adding a formerly backing index to a data stream?

[the remark this answers is not in the document]
Comment (danhermann) [reply server/src/test/java/org/elasticsearch/cluster/metadata/MetadataDataStreamsServiceTests.java:354]: Yes, it's this:

```
org.elasticsearch.cluster.metadata.MetadataDataStreamsServiceTests > testAddRemoveAddRoundtripInSingleRequest FAILED
    org.elasticsearch.index.mapper.MapperParsingException: Failed to parse mapping: Root mapping definition has unsupported parameters:  [_doc : {_data_stream_timestamp={enabled=true}, properties={@timestamp={type=date}}}]
        at __randomizedtesting.SeedInfo.seed([7EDB2DEA38BEEF2B:7C913494D2242507]:0)
        at app//org.elasticsearch.index.mapper.MapperService.parseMapping(MapperService.java:303)
        at app//org.elasticsearch.index.mapper.MapperService.mergeAndApplyMappings(MapperService.java:280)
        at app//org.elasticsearch.index.mapper.MapperService.merge(MapperService.java:271)
        at app//org.elasticsearch.index.mapper.MapperServiceTestCase.merge(MapperServiceTestCase.java:264)
        at app//org.elasticsearch.index.mapper.MapperServiceTestCase.createMapperService(MapperServiceTestCase.java:146)
        at app//org.elasticsearch.cluster.metadata.MetadataDataStreamsServiceTests.getMapperService(MetadataDataStreamsServiceTests.java:350)
```

[the remark this answers is not in the document]
Comment (martijnvg) [reply server/src/test/java/org/elasticsearch/cluster/metadata/MetadataDataStreamsServiceTests.java:354]: Ok, looks like it doesn't know about the custom data stream metadata field added via the `MetadataIndexTemplateServiceTests.DummyPlugin`. Let me take a better look.

**Comment by danhermann** [comment]
> Ok, looks like it doesn't know about the custom data stream metadata field added via the `MetadataIndexTemplateServiceTests.DummyPlugin`. Let me take a better look.

@martijnvg, were you able to find anything about the plugin and the metadata field?

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by martijnvg** [comment]
> were you able to find anything about the plugin and the metadata field?

Yes, I debugged this and for the failing tests the mapping got wrapped twice in a `_doc` json object. 
I pushed a fix, which I think should make the tests pass 🤞.

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
> > were you able to find anything about the plugin and the metadata field?
> 
> Yes, I debugged this and for the failing tests the mapping got wrapped twice in a `_doc` json object. I pushed a fix, which I think should make the tests pass 🤞.

Thank you, @martijnvg! I think this is ready for review, then.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDataStreamsService.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    private final IndicesService indicesService;
+
+    public MetadataDataStreamsService(ClusterService clusterService, IndicesService indicesService) {
+        this.clusterService = clusterService;
+        this.indicesService = indicesService;
+    }
+
+    public void updateBackingIndices(final ModifyDataStreamRequest request,
+                                     final ActionListener<AcknowledgedResponse> listener) {
+        clusterService.submitStateUpdateTask("update-backing-indices",
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDataStreamsService.java:42]: Maybe add a check that checks whether there are any actions in the provided request and if not immediately invoke the listener's onReponse method?

**Review by martijnvg — APPROVED** [review]
LGTM

**Comment by danhermann** [comment]
Thanks for the review, @martijnvg!

**Pull request merged**